### PR TITLE
Correct i18n-ellipsis fix quotes

### LINKED ISSUE #44: i18n-ellipsis: Fixer removes quotes
When running `eslint --fix` on a rule which violates the `wpcalypso/i18n-ellipsis` rule, the enclosing quotes are removed from the string.

```js
this.props.translate( 'Importing...' );

// run `eslint --fix` on offending file

this.props.translate( Importing… );  // 😱 😱 😱 
```

Context: I came across this while codemodding Calypso (https://github.com/Automattic/wp-calypso/pull/18263)

### PR DESCRIPTION
**Pull request by sirreal**
Adds test for `i18n-ellipsis` fix.
Wraps fixed `i18n-ellipsis` in quotes.

~This fix continues to have problems with strings and interpolation:~ I believe this is tested and fixed now.

```js
translate( `Hello ${ "World" }...` );

// becomes

translate( 'Hello …' );
```

I believe there is another lint rule for use of variables within translated strings. If `translate` is being used correctly this shouldn't be an issue, but that's what the linter is for, isn't it 🙂 

~I'd like to find a more robust solution~ 🤔  I think we're good now 👍 

Fixes #44

### CODE REVIEW

**Comment by sirreal** [comment]
This may have gotten a little out of hand 😅 

I started running into edge cases, most of which should _never_ be used in a translated string. Other rules should throw errors if they do not already.

I've added several tests for edge cases, and improved handling of some known cases. Most importantly, the `Literal` case should be fixed by maintaining the original quotes.

**Comment by aduth** [comment]
Is template literal something we'd even want to encourage or support for `translate` calls? Seems like it'd verge on encouraging variable interpolation which we [otherwise disallow](https://github.com/Automattic/eslint-plugin-wpcalypso/blob/master/docs/rules/i18n-no-variables.md).

**Comment by sirreal** [comment]
> Is template literal something we'd even want to encourage or support for translate calls?

I don't think it should be encouraged. I'd speculate it's useful for multi-line strings, but variables and interpolation clearly should be avoided.

That said, my goal here is to avoid producing bad output when `eslint --fix` is run, which is why I've tried to cover a reasonable set of inputs.

I believe this is the only rule that actually provides a fix in this repo. If we're concerned about it's behavior I'd be fine with just removing the `fix` altogether.

File: lib/rules/i18n-ellipsis.js [hunk 1]
```diff
@@ -26,12 +38,37 @@ var rule = module.exports = function( context ) {
...
-				if ( argumentString && -1 !== argumentString.indexOf( '...' ) ) {
+				if ( argumentString && containsThreeDots( argumentString ) ) {
 					context.report( {
 						node: arg,
 						message: rule.ERROR_MESSAGE,
 						fix: function( fixer ) {
-							return fixer.replaceText( arg, argumentString.replace( /\.\.\./g, '…' ) );
+							if ( arg.type === 'TemplateLiteral' ) {
+								var fixes = [];
+								arg.quasis.forEach(
```
Comment (aduth) [inline lib/rules/i18n-ellipsis.js]: Minor: Could accumulate this with `Array#reduce`

```js
return arg.quasis.reduce( ( fixes, quasi ) => {
	if ( /* ... */ ) {
		fixes.push( /* ... */ );
	}

	return result;
}, [] );

File: lib/rules/i18n-ellipsis.js [hunk 2]
```diff
@@ -26,12 +38,37 @@ var rule = module.exports = function( context ) {
...
+										}
+									}
+								);
+								return fixes;
+							}
+							if ( arg.type === 'Literal' ) {
+								return fixer.replaceText( arg, replaceThreeDotsWithEllipsis( arg.raw ) );
+							}
+							return fixer.replaceText( arg,
+								"'" +
```
Comment (aduth) [inline lib/rules/i18n-ellipsis.js]: They really don't make it easy to replace text of a BinaryExpression, huh? 😄 

That's what's being handled here, right? Should we be explicit about that? Maybe worth a `switch` on `arg.type` ?

**Review by aduth — COMMENTED** [review]
Nicely written tests. This looks pretty good to me, despite the necessary complexity to implement.

File: lib/rules/i18n-ellipsis.js (hunk 1 above)
Comment (sirreal) [reply lib/rules/i18n-ellipsis.js]: ❤️  nice, `reduce` in 0638c7cb7488624725d0348e3795cfe063704df7

File: lib/rules/i18n-ellipsis.js (hunk 2 above)
Comment (sirreal) [reply lib/rules/i18n-ellipsis.js]: This is the previous implementation corrected to wrap in single quotes and escape contained single quotes.

Initially, I didn't go any further because I wasn't sure at what point to stop implementing handling for different types expressions. 

Since you've called my attention to `BinaryExpression`, I realize that that is a natural stopping point because it's the last case covered by `getTextContentFromNode`. Anything not covered by those three cases should not throw a lint error because the text won't be included.

I've changed the implementation in f0db405 so that we can recursively fix the left and right sides of `BinaryExpression`s. I'm quite satisfied with the implementation at this point.

Note that since the fixer now works "natively" with eslint functionality and doesn't rely on `getTextContentFromNode` to manually modify the string contents, the problems with quotes vanish. `+` joined strings and different quote types are untouched. Compare that to the initial implementation which mangled the source quite badly when it attempted to fix:

https://github.com/Automattic/eslint-plugin-wpcalypso/pull/45/files#diff-72e7d21b11fe78ad24318057d029a272R97

**Review by aduth — APPROVED** [review]
Recent revisions look great 👍  Recursing into BinaryExpression appears to work much better. I'll plan to include this in a patch release.

**Comment by sirreal** [comment]
Thanks @aduth! What's the policy here, shall I go ahead and merge this PR or let you handle it?

**Comment by aduth** [comment]
Sure, it's fine to merge. I'll update the `package.json`, `package-lock.json`, and `CHANGELOG.md` files on master.

**Comment by aduth** [comment]
Thanks for the fix, and for the patience in my delayed reviews 😄 

Published to npm as v4.0.1
